File: frontend/app/filmConfig.ts
export const films = [
  { slug: "shawshank-redemption", title: "The Shawshank Redemption", lenses: ["Freedom", "Hope", "Institutional Control", "Friendship", "Justice"] },
  { slug: "fight-club", title: "Fight Club", lenses: ["Identity", "Masculinity", "Consumerism", "Violence", "Doubles"] },
  { slug: "one-flew-over-the-cuckoos-nest", title: "One Flew Over the Cuckoo's Nest", lenses: ["Control", "Institutional Power", "Freedom", "Rebellion", "Madness"] },
  { slug: "se7en", title: "Se7en", lenses: ["Justice", "Violence", "Moral Decay", "Obsession", "Guilt"] },
  { slug: "silence-of-the-lambs", title: "The Silence of the Lambs", lenses: ["Power", "Fear", "Identity", "Gender", "Control"] },
  { slug: "the-prestige", title: "The Prestige", lenses: ["Obsession", "Performance", "Sacrifice", "Doubles", "Truth"] },
  { slug: "memento", title: "Memento", lenses: ["Memory", "Truth", "Identity", "Guilt", "Self-Deception"] },
  { slug: "taxi-driver", title: "Taxi Driver", lenses: ["Isolation", "Masculinity", "Violence", "Alienation", "Moral Delusion"] },
  { slug: "shutter-island", title: "Shutter Island", lenses: ["Reality vs Illusion", "Trauma", "Guilt", "Denial", "Madness"] },
  { slug: "black-swan", title: "Black Swan", lenses: ["Performance", "Identity", "Obsession", "Control", "Doubles"] },
  { slug: "sixth-sense", title: "The Sixth Sense", lenses: ["Grief", "Perception", "Denial", "Childhood", "Revelation"] },
  { slug: "prisoners", title: "Prisoners", lenses: ["Justice", "Faith", "Violence", "Obsession", "Moral Ambiguity"] },
  { slug: "gone-girl", title: "Gone Girl", lenses: ["Performance", "Marriage", "Media", "Control", "Identity"] },
  { slug: "requiem-for-a-dream", title: "Requiem for a Dream", lenses: ["Addiction", "Obsession", "Desire", "Decay", "Control"] },
  { slug: "donnie-darko", title: "Donnie Darko", lenses: ["Time", "Fate", "Reality vs Illusion", "Alienation", "Madness"] },
  { slug: "the-machinist", title: "The Machinist", lenses: ["Guilt", "Insomnia", "Body", "Madness", "Self-Punishment"] },
  { slug: "mulholland-drive", title: "Mulholland Drive", lenses: ["Dream Logic", "Identity", "Desire", "Hollywood", "Reality vs Illusion"] },
  { slug: "truman-show", title: "The Truman Show", lenses: ["Surveillance", "Freedom", "Reality vs Illusion", "Control", "Performance"] },
] as const;

export const allLenses: string[] = Array.from(new Set(films.flatMap((film) => film.lenses))).sort();

export const demos = [
  { mode: "analyze_film", filmA: "memento", lens: "Memory", label: "Memento", helper: "Memory" },
  { mode: "analyze_film", filmA: "mulholland-drive", lens: "Reality vs Illusion", label: "Mulholland Drive", helper: "Reality vs Illusion" },
  { mode: "analyze_film", filmA: "black-swan", lens: "Performance", label: "Black Swan", helper: "Performance" },
  { mode: "analyze_film", filmA: "taxi-driver", lens: "Isolation", label: "Taxi Driver", helper: "Isolation" },
  { mode: "analyze_film", filmA: "truman-show", lens: "Control", label: "The Truman Show", helper: "Control" },
] as const;

File: backend/app/film_config.py
"""Film catalog and dynamic lens-query expansion.
"""

from __future__ import annotations

import re

from app.services.lens_profiles import selected_profile_terms


FILMS = [
    {"slug": "shawshank-redemption", "title": "The Shawshank Redemption"},
    {"slug": "fight-club", "title": "Fight Club"},
    {"slug": "one-flew-over-the-cuckoos-nest", "title": "One Flew Over the Cuckoo's Nest"},
    {"slug": "se7en", "title": "Se7en"},
    {"slug": "silence-of-the-lambs", "title": "The Silence of the Lambs"},
    {"slug": "the-prestige", "title": "The Prestige"},
    {"slug": "memento", "title": "Memento"},
    {"slug": "taxi-driver", "title": "Taxi Driver"},
    {"slug": "shutter-island", "title": "Shutter Island"},
    {"slug": "black-swan", "title": "Black Swan"},
    {"slug": "sixth-sense", "title": "The Sixth Sense"},
    {"slug": "prisoners", "title": "Prisoners"},
    {"slug": "gone-girl", "title": "Gone Girl"},
    {"slug": "requiem-for-a-dream", "title": "Requiem for a Dream"},
    {"slug": "donnie-darko", "title": "Donnie Darko"},
    {"slug": "the-machinist", "title": "The Machinist"},
    {"slug": "mulholland-drive", "title": "Mulholland Drive"},
    {"slug": "truman-show", "title": "The Truman Show"},
]

FILM_TITLES = {film["slug"]: film["title"] for film in FILMS}


def _terms(value: str | None) -> list[str]:
    if not value:
        return []
    return [value, *re.findall(r"[A-Za-z][A-Za-z'-]+", value)]


def _dedupe(terms: list[str]) -> list[str]:
    seen: set[str] = set()
    result: list[str] = []
    for term in terms:
        normalized = str(term).strip()
        if normalized and normalized.lower() not in seen:
            seen.add(normalized.lower())
            result.append(normalized)
    return result


def expand_lens_terms(lens: str | None) -> list[str]:
    """Return the user-selected direct lens without a static vocabulary."""
    return _dedupe(_terms(lens))


def expand_film_lens_terms(film_slug: str | None, lens: str | None) -> list[str]:
    """Expand only from the evidence-backed direct profile definition."""
    if not lens:
        return []
    values = selected_profile_terms(film_slug, lens) if film_slug else [lens]
    return _dedupe([term for value in values for term in _terms(value)])
